src/runtime/manifest-parser.pegjs

```diff
@@ -1581,6 +1584,84 @@ SchemaPrimitiveType
     });
   }
 
+Adapter "an adapter, (e.g. adapter Foo(param: Person { name: Text }) => Friend { nickName: param.name } )"
+  = 'adapter' whiteSpace adapterName:AdapterName whiteSpace? '(' multiLineSpace? params:AdapterParamsDeclaration multiLineSpace? ')' whiteSpace? '=>' multiLineSpace? body:AdapterBodyDefinition eolWhiteSpace
+  {
+     return toAstNode<AstNode.AdapterNode>({
+       kind: 'adapter-node',
+       name: adapterName,
+       params,
+       body
+     });
+  }
+
+AdapterName
+  = upperIdent
+
+AdapterParamName
+  = fieldName
+
+AdapterParamsDeclaration
+  = param:AdapterParam restParams:(whiteSpace? ',' multiLineSpace AdapterParam)* {
+     const params = [param].concat(restParams.map(rparam => rparam[3]));
+     const names = params.map(p => p.name);
+     const duplicateNames = names.filter((item, index) => names.indexOf(item) != index);
+     if (duplicateNames.length) {
+        error(`Duplicate adapter param names ${duplicateNames.join(',')}`);
+     } else {
+        return params;
+     }
+  }
+
+AdapterParam
+  = paramName:AdapterParamName whiteSpace? ':' whiteSpace? type:ParticleHandleConnectionType {
+     return toAstNode<AstNode.AdapterParam>({
+        kind: 'adapter-param',
+        name: paramName,
+        type
+     });
+  }
+
+AdapterBodyDefinition
+  = name:('*' / upperIdent) whiteSpace? '{' multiLineSpace fields:AdapterFields? ','? multiLineSpace '}' {
```

Comment:
Nice, we need to add an option of an optional trailing comma to schemas as well!